### PR TITLE
SNOW-642635: Remove legacy ocsp cache code

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-642635

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [x] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

- Remove SSD code which is never used
- Refactor the OCSP_CACHE, consolidating the old design with the new design:
  - one in-memory cache: OCSP_RESPONSE_VALIDATION_CACHE, which caches not only ocsp response, but also the validation result
  - two files: OCSP_RESPONSE_VALIDATION_CACHE which is the dump of the in-memory one, and OCSP_RESPONSE which stores the raw ocsp responses in the same format as downloaded from the snowflake cache service.

### CODE REVIEW

File: src/snowflake/connector/ocsp_asn1crypto.py [hunk 1]
```diff
@@ -104,11 +97,7 @@ def encode_cert_id_key(self, hkey):
         return cert_id
 
     def decode_cert_id_key(self, cert_id):
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_asn1crypto.py:99]: type-hints
```suggestion
    def decode_cert_id_key(self, cert_id: CertId) -> tuple[bytes, bytes, bytes]:
```

File: src/snowflake/connector/ocsp_snowflake.py [hunk 2]
```diff
@@ -59,39 +56,57 @@
...
         },
     )
 except OSError:
     # In case we run into some read/write permission error fall back onto
     #  in memory caching
-    OCSP_CACHE: SFDictCache[
-        tuple[bytes, bytes, int, str],
-        tuple[Exception | None, Certificate, Certificate, CertId, bytes],
+    OCSP_RESPONSE_VALIDATION_CACHE: SFDictCache[
+        tuple[bytes, bytes, int],
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:108]: You modified [generate_cache_key's signature](https://github.com/snowflakedb/snowflake-connector-python/blob/3a94c4a1c07f460e07d09f051e0cd4725a7b51e0/src/snowflake/connector/ocsp_snowflake.py#L119), based on that I believe this should be
```suggestion
        tuple[bytes, bytes, bytes]
```

File: src/snowflake/connector/ocsp_snowflake.py [hunk 3]
```diff
@@ -717,43 +703,27 @@ def is_cache_fresh(current_time, ts):
...
 
     @staticmethod
-    def find_cache(ocsp, cert_id, subject):
+    def find_cache(ocsp, cert_id, subject, **kwargs):
         subject_name = ocsp.subject_name(subject) if subject else None
         current_time = int(time.time())
-        hkey = ocsp.decode_cert_id_key(cert_id)
-        if hkey in OCSPCache.CACHE:
-            ts, cache = OCSPCache.CACHE[hkey]
+        cache_key = kwargs.get("cache_key", ocsp.decode_cert_id_key(cert_id))
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:709]: Please type-hint this line.
Maybe?
```suggestion
        cache_key: tuple[bytes, bytes, bytes] = kwargs.get("cache_key", ocsp.decode_cert_id_key(cert_id))
```
But that'd mean that my comment about `OCSP_RESPONSE_VALIDATION_CACHE`'s type-hint was correct

File: src/snowflake/connector/ocsp_snowflake.py [hunk 4]
```diff
@@ -775,58 +745,27 @@ def update_or_delete_cache(ocsp, cert_id, ocsp_response, ts):
...
-    def delete_cache(ocsp, cert_id):
-        try:
-            with OCSPCache.CACHE_LOCK:
-                hkey = ocsp.decode_cert_id_key(cert_id)
-                if hkey in OCSPCache.CACHE:
-                    del OCSPCache.CACHE[hkey]
-                    OCSPCache.CACHE_UPDATED = True
-        except Exception as ex:
-            logger.debug("Could not acquire lock", ex)
+        cache_key = kwargs.get("cache_key", ocsp.decode_cert_id_key(cert_id))
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:751]: Type-hint this line, please

File: src/snowflake/connector/ocsp_snowflake.py [hunk 5]
```diff
@@ -59,39 +56,57 @@
...
+class OCSPResponseValidationMetadata(NamedTuple):
+    exception: Exception | None = None
+    issuer: Certificate | None = None
+    subject: Certificate | None = None
+    cert_id: CertId | None = None
+    ocsp_response: bytes | None = None
+
+
+class OCSPResponseValidationResult(NamedTuple):
+    ocsp_response_validation_metadata: OCSPResponseValidationMetadata | None = None
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:76]: Do you think that we could flatten the 2 NamedTuples into just one?

File: src/snowflake/connector/ocsp_snowflake.py [hunk 6]
```diff
@@ -775,58 +745,27 @@ def update_or_delete_cache(ocsp, cert_id, ocsp_response, ts):
...
     @staticmethod
-    def merge_cache(
-        ocsp, previous_cache_filename, current_cache_filename, output_filename
-    ):
-        """Merges two cache files into one cache and save to the output.
-
-        current_cache takes precedence over previous_cache.
-        """
+    def delete_cache(ocsp, cert_id, **kwargs):
+        cache_key = kwargs.get("cache_key", ocsp.decode_cert_id_key(cert_id))
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:763]: Type-hint

File: src/snowflake/connector/ocsp_snowflake.py [hunk 7]
```diff
@@ -775,58 +745,27 @@ def update_or_delete_cache(ocsp, cert_id, ocsp_response, ts):
...
 
     @staticmethod
-    def merge_cache(
-        ocsp, previous_cache_filename, current_cache_filename, output_filename
-    ):
-        """Merges two cache files into one cache and save to the output.
-
-        current_cache takes precedence over previous_cache.
-        """
+    def delete_cache(ocsp, cert_id, **kwargs):
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:762]: Type-hint

File: src/snowflake/connector/ocsp_snowflake.py [hunk 8]
```diff
@@ -1237,58 +1088,14 @@ def validate_by_direct_connection(
         telemetry_data: OCSPTelemetryData,
         hostname: str = None,
         do_retry: bool = True,
+        **kwargs,
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:1091]: I like to add explicit Anys (even it they are implicitly there already) to show that typing was added and not forgotten.
```suggestion
        **kwargs: Any,
```

File: src/snowflake/connector/ocsp_snowflake.py [hunk 9]
```diff
@@ -1187,7 +1036,7 @@ def _validate(
     def get_ocsp_retry_choice():
         return os.getenv("SF_OCSP_DO_RETRY", "true") == "true"
 
-    def is_cert_id_in_cache(self, cert_id, subject):
+    def is_cert_id_in_cache(self, cert_id, subject, **kwargs):
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:1039]: Type-hint

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: src/snowflake/connector/cache.py [hunk 10]
```diff
@@ -332,6 +336,9 @@ def _add_or_remove(self) -> None:
         """
         self.telemetry["size"] = len(self._cache)
 
+    def _save(self, load_first: bool = True) -> bool:
+        pass
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py]: good point, updated to remove this method from SFDictCache

File: src/snowflake/connector/ocsp_snowflake.py [hunk 11]
```diff
@@ -59,39 +56,57 @@
...
 from .cache import SFDictCache, SFDictFileCache
 from .telemetry import TelemetryField, generate_telemetry_data_dict
 
+
+class OCSPResponseValidationMetadata(NamedTuple):
+    exception: Exception | None = None
+    issuer: Certificate | None = None
+    subject: Certificate | None = None
+    cert_id: CertId | None = None
+    ocsp_response: bytes | None = None
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py:72]: when first time loading from the file ocsp_response.json, ocsp_response hasn't been validated yet, so can't fill in the fields until validated

File: src/snowflake/connector/ocsp_snowflake.py [hunk 12]
```diff
@@ -349,30 +346,18 @@ def reset_ocsp_dynamic_cache_server_url(self, use_ocsp_cache_server):
...
-                        self.OCSP_RETRY_URL = (
-                            "{}://{}/retry/".format(
-                                parsed_url.scheme, parsed_url.hostname
-                            )
-                            + "{0}/{1}"
-                        )
-                else:
-                    if parsed_url.port:
-                        self.OCSP_RETRY_URL = "{}://{}:{}/retry".format(
+                if parsed_url.port:
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: updated to use `urlunparse`, code is greatly simplified, thanks for the suggestion

File: src/snowflake/connector/ocsp_snowflake.py [hunk 13]
```diff
@@ -349,30 +346,18 @@ def reset_ocsp_dynamic_cache_server_url(self, use_ocsp_cache_server):
...
+                    self.OCSP_RETRY_URL = (
+                        "{}://{}:{}/retry/".format(
                             parsed_url.scheme, parsed_url.hostname, parsed_url.port
                         )
-                    else:
-                        self.OCSP_RETRY_URL = "{}://{}/retry".format(
-                            parsed_url.scheme, parsed_url.hostname
-                        )
+                        + "{0}/{1}"
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: I don't think it's part of the SSD path -- it landed in the section `if not OCSPCache.ACTIVATE_SSD:` and  it's for retry downloading the ocsp response json file and

File: src/snowflake/connector/ocsp_snowflake.py [hunk 14]
```diff
@@ -349,30 +346,18 @@ def reset_ocsp_dynamic_cache_server_url(self, use_ocsp_cache_server):
...
+                    )
+                else:
+                    self.OCSP_RETRY_URL = (
+                        f"{parsed_url.scheme}://{parsed_url.hostname}/retry/"
+                        + "{0}/{1}"
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: same as above

File: src/snowflake/connector/ocsp_snowflake.py [hunk 15]
```diff
@@ -391,7 +376,9 @@ def download_cache_from_server(self, ocsp):
                 logger.debug(
                     "downloaded OCSP response cache file from %s", self.CACHE_SERVER_URL
                 )
-                logger.debug("# of certificates: %s", len(OCSPCache.CACHE))
+                logger.debug(
+                    "# of certificates: %s", len(OCSP_RESPONSE_VALIDATION_CACHE)
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: definitely agree upon it, changed to use `._cache` to unblock the logic.

File: src/snowflake/connector/ocsp_snowflake.py [hunk 16]
```diff
@@ -555,7 +534,9 @@ def reset_ocsp_response_cache_uri(ocsp_response_cache_uri):
             )
 
         logger.debug("ocsp_response_cache_uri: %s", OCSPCache.OCSP_RESPONSE_CACHE_URI)
-        logger.debug("OCSP_VALIDATION_CACHE size: %s", len(OCSPCache.CACHE))
+        logger.debug(
+            "OCSP_VALIDATION_CACHE size: %s", len(OCSP_RESPONSE_VALIDATION_CACHE)
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: updated

File: src/snowflake/connector/ocsp_snowflake.py [hunk 17]
```diff
@@ -593,7 +574,7 @@ def read_ocsp_response_cache_file(ocsp, filename):
                 logger.debug(
                     "Read OCSP response cache file: %s, count=%s",
                     filename,
-                    len(OCSPCache.CACHE),
+                    len(OCSP_RESPONSE_VALIDATION_CACHE),
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: updated

File: src/snowflake/connector/ocsp_snowflake.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: fixed

File: src/snowflake/connector/ocsp_snowflake.py (hunk 4 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: updated

File: src/snowflake/connector/ocsp_snowflake.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: you're correct, updated

File: src/snowflake/connector/ocsp_snowflake.py (hunk 6 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: updated

File: src/snowflake/connector/ocsp_snowflake.py (hunk 7 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: updated

File: src/snowflake/connector/ocsp_snowflake.py (hunk 5 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: yes, updated to just use one

File: src/snowflake/connector/ocsp_snowflake.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sfc-gh-mkeller) [reply src/snowflake/connector/ocsp_snowflake.py]: Oops, I missed the `not`

**Review by sfc-gh-mkeller — APPROVED** [review]
:ship: :ship:

**Pull request merged**